src/_pytest/terminal.py

```diff
@@ -483,43 +484,51 @@ def pytest_runtest_logreport(self, report: TestReport) -> None:
 
     @property
     def _is_last_item(self):
-        return len(self._progress_nodeids_reported) == self._session.testscollected
+        return self._progress_items_reported == self._session.testscollected
 
-    def pytest_runtest_logfinish(self, nodeid):
-        assert self._session
-        if self.verbosity <= 0 and self._show_progress_info:
-            if self._show_progress_info == "count":
-                num_tests = self._session.testscollected
-                progress_length = len(" [{}/{}]".format(str(num_tests), str(num_tests)))
-            else:
-                progress_length = len(" [100%]")
+    def pytest_runtest_teardown(self) -> None:
+        """Write progress if past edge."""
+        if self.verbosity >= 0 or not self._show_progress_info:
+            return
 
-            self._progress_nodeids_reported.add(nodeid)
+        if self._show_progress_info == "count":
+            assert self._session
+            num_tests = self._session.testscollected
+            progress_length = len(" [{0}/{0}]".format(num_tests))
+        else:
+            progress_length = len(" [100%]")
 
-            if self._is_last_item:
-                self._write_progress_information_filling_space()
-            else:
-                main_color, _ = self._get_main_color()
-                w = self._width_of_current_line
-                past_edge = w + progress_length + 1 >= self._screen_width
-                if past_edge:
-                    msg = self._get_progress_information_message()
-                    self._tw.write(msg + "\n", **{main_color: True})
+        w = self._width_of_current_line
+        past_edge = w + progress_length + 1 >= self._screen_width
+        if past_edge:
+            msg = self._get_progress_information_message()
+            main_color, _ = self._get_main_color()
+            self._tw.write(msg + "\n", **{main_color: True})
+
+    @pytest.hookimpl(hookwrapper=True)
+    def pytest_runtestloop(self) -> Generator[None, None, None]:
+        """Write final progress indicator."""
+        yield
+        if (
+            getattr(self, "_tests_ran", False)
```

Comment:
I think we should avoid attributes being set but not declared on `__init__` when that is possible; this has been a great source of confusion/bugs in the past.